### DISCUSSION TITLE
Does cutlass_profiler support f8 gemm?

### DISCUSSION THREAD

**Discussion by liligwu** [NVIDIA/cutlass#1217] [General]
As 3.2 supports f8 gemm, does the profiler cutlass_profiler support it as well? Thanks. 
I tested like 
`./cutlass_profiler --operation=Gemm --accum=f32 --m=16 --n=16 --k=1024 --A=f16:* --B=f8:* --C=f16:* --providers=cutlass --output=gemm_out.csv`. It's not working. And I couldn't find any f8-related code in the profiler.

**Comment by thakkarV** [comment]
Yes it does. There are two f8 types, e4m3 and e5m2. Look for those specifically and you should be able to find them

**Reply by liligwu** [reply]
Hi @thakkarV, thanks for your quick reply. Would you please give an example of running command using f8 in `cutlass_profiler`?

**Reply by thakkarV** [reply]
cutlass_profiler --kernels=cutlass3x_sm90_tensorop_s64x128x32gemm_e5m2_e4m3_f32_f32_f32_128x128x128_2x1x1_0_tnn_align16_stream_k_warpspecialized_cooperative_epi_nosmem --m=2048 --n=8448 --k=8192

**Reply by thakkarV** [reply]
or 

```
./cutlass_profiler --operation=Gemm --accum=f32 --m=16 --n=16 --k=1024 --A=e4m3:* --B=em52:* --C=f32:* --providers=cutlass --output=gemm_out.csv
```

**Comment by liligwu** [comment]
Hi @thakkarV, Thanks for your examples. Currently, I have a trouble of building cutlass_profiler on H100. The following dockerfile crashes the system.

```
ARG BASE_DOCKER=nvcr.io/nvidia/pytorch:23.07-py3
FROM $BASE_DOCKER
USER root

ENV WORKSPACE_DIR=/workspace
RUN mkdir -p $WORKSPACE_DIR
WORKDIR $WORKSPACE_DIR

RUN git clone https://github.com/NVIDIA/cutlass.git \
    && cd cutlass \
    && mkdir build && cd build \
    && cmake .. -DCUTLASS_NVCC_ARCHS=90a \
    && make cutlass_profiler -j \
    && cd examples \
    && make -j
```
The system has two H100-PCIe. I'm attaching the system info and nvidia-smi. I also attempted to cross-compile it on a V100, it crashed the V100 system as well. 
[nvidia-smi.txt](https://github.com/NVIDIA/cutlass/files/13619487/nvidia-smi.txt)
[config.txt](https://github.com/NVIDIA/cutlass/files/13619488/config.txt)

Please let me know if you can reproduce the issue. Thank you.

**Reply by thakkarV** [reply]
do you have any relevant logs or error messages? I don't see anything indicating a crash in the config or the nvidia-smi output that you have included. the command list does not look out of the ordinary for me. Are you able to build without docker ?

**Reply by liligwu** [reply]
The build stops at around 93% after `make cutlass_profiler -j`, and I simply lose the connection to the system. I have to power cycle it. I can try it without docker, but it would be good if it works in a docker for us. Thanks.

**Reply by liligwu** [reply]
I tried nvcr.io/nvidia/pytorch:23.11-py3. The same outcome.

**Reply by thakkarV** [reply]
its hard for me to help without knowing what the failure cause is, and since I do not have a docker setup, its not possible for me to repro this either. Adding @jackkosaian since he has a pytorch docker setup though I believe.

**Comment by d-k-b** [comment]
Can you try with `-j 8` or smaller set of parallel processes. Likely the
image is hanging due to resource contention.

**Reply by liligwu** [reply]
Reducing the MAX_JOB to 8 works. Thanks @d-k-b , @thakkarV

**Comment by liligwu** [comment]
Hi @thakkarV , sorry to bother you again. I built the docker successfully, but still have problems running the profiler on f8. specifically, 
`./cutlass_profiler --operation=Gemm --accum=f32 --m=16 --n=16 --k=1024 --A=f16:* --B=f16:* --C=f16:* --providers=cutlass --output=gemm_out.csv` 
returns a correct csv file. but 
`./cutlass_profiler --operation=Gemm --accum=f32 --m=16 --n=16 --k=1024 --A=f16:* --B=e5m2:* --C=f16:* --providers=cutlass --output=gemm_out.csv` returns a csv file with header only. 
Did I miss anything? Thank you.

**Reply by thakkarV** [reply]
Your A is f16 but B is e5m2 in the command line above. There is no kernel matching that description in the profiler. Both A and B must be an f8 type.

**Reply by liligwu** [reply]
I see. but `./cutlass_profiler --operation=Gemm  --m=16 --n=16 --k=1024 --A=e5m2:* --B=e5m2:* --C=f16:* --providers=cutlass --o
utput=gemm_out.csv` returns an empty csv file as well.

**Reply by liligwu** [reply]
Should I add `-DCUTLASS_LIBRARY_KERNELS=all  -DCUTLASS_UNITY_BUILD_ENABLED=ON` when make the profiler?

**Reply by thakkarV** [reply]
Yes. Please also try fp32 C type.

**Reply by liligwu** [reply]
Thanks, I'll try it out and let you know.

**Reply by liligwu** [reply]
Hi @thakkarV , currently, I'm able to run `cutlass_profiler --kernels=cutlass3x_sm90_tensorop_s64x128x32gemm_e5m2_e4m3_f32_f32_f32_128x128x128_2x1x1_0_tnn_align16_stream_k_warpspecialized_cooperative_epi_nosmem --m=2048 --n=8448 --k=8192` in the container, but `./cutlass_profiler --operation=Gemm  --m=16 --n=16 --k=1024 --A=e5m
2:* --B=e5m2:* --C=f32:* --providers=cutlass --output=gemm_out.csv` still failed. Would you happen to have any suggestions? We prefer the latter format as we don't care which kernel is invoked. Thanks.

**Reply by thakkarV** [reply]
What does failure mean here?

**Reply by liligwu** [reply]
There are no failure messages, it outputs an empty .csv file.

**Reply by thakkarV** [reply]
`--A=e5m 2:*` looks like you have a spurious space here.

**Reply by liligwu** [reply]
Sorry, it was a copy-past issue, but the command outputs a .csv file with a header only even without this blank space.

**Reply by liligwu** [reply]
Hi @thakkarV , I'd appreciate it if you could provide any suggestions. Thank you.

**Reply by thakkarV** [reply]
Output type you are selecting is fp16 (`--C=f16:*`), but are you stamping out any kernels with fp16 as the output type for fp8 inputs?

**Reply by liligwu** [reply]
Hi @thakkarV , I used `-DCUTLASS_LIBRARY_KERNELS=all  -DCUTLASS_UNITY_BUILD_ENABLED=ON` when building the profiler.

**Reply by liligwu** [reply]
Hi @thakkarV , do you have any examples or documents available for `./cutlass_profiler --operation=Gemm` format? Thank you.

**Reply by thakkarV** [reply]
I don't think we have any OOTB f16 output kernels in the profiler for fp8. you can add them to the generator yourself and then set C to fp16, or switch to using fp32 output. The profiler documentation is here: https://github.com/NVIDIA/cutlass/blob/main/media/docs/profiler.md

**Reply by liligwu** [reply]
Yeah, it doesn't matter about the output precision. Would you please give a legitimate example of fp8 profiling command in `./cutlass_profiler --operation=Gemm` format? Thanks.

**Reply by thakkarV** [reply]
`./tools/profiler/cutlass_profiler --m=16 --n=16 --k=1024 --A=fe5m2:\* --B=fe5m2:\*`

works for me just fine, or any other combination of fp8 types, layouts etc. I also noticed that your A type if fp16 but B is fp8. Mixed input gemms are supported in CUTLASS but not plumbed into the profiler yet.

**Reply by liligwu** [reply]
@thakkarV , this command finally works. Thank you very much!

**Comment by hjq0127** [comment]
Hello. Could you please tell me the actual FP8 test value for the H100? I'm using Cutlass instance 54 for testing, but it consistently fails to meet the standard.
<img width="1282" height="708" alt="1741776238652_ pic_hd" src="https://github.com/user-attachments/assets/ab455509-4ab1-4dca-a46d-d7e35149c274" />
